File: openapi/openapi-generator/src/main/java/ru/tinkoff/kora/openapi/generator/javagen/ServerApiModuleGenerator.java
package ru.tinkoff.kora.openapi.generator.javagen;

import com.palantir.javapoet.ClassName;
import com.palantir.javapoet.JavaFile;
import com.palantir.javapoet.MethodSpec;
import com.palantir.javapoet.TypeSpec;
import org.openapitools.codegen.model.OperationsMap;
import ru.tinkoff.kora.openapi.generator.DelegateMethodBodyMode;

import javax.lang.model.element.Modifier;

public class ServerApiModuleGenerator extends AbstractJavaGenerator<OperationsMap> {
    @Override
    public JavaFile generate(OperationsMap ctx) {
        var b = TypeSpec.interfaceBuilder(ctx.get("classname") + "Module")
            .addModifiers(Modifier.PUBLIC)
            .addAnnotation(generated())
            .addAnnotation(Classes.module);
        if (params.delegateMethodBodyMode != DelegateMethodBodyMode.NONE) {
            var delegateClass = ClassName.get(apiPackage, ctx.get("classname") + "Delegate");
            b.addMethod(MethodSpec.methodBuilder("default" + ctx.get("classname") + "Delegate")
                .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
                .addStatement("return new $T() {}", delegateClass)
                .returns(delegateClass)
                .build());
        }

        return JavaFile.builder(apiPackage, b.build()).build();

    }
}

File: openapi/openapi-generator/src/main/kotlin/ru/tinkoff/kora/openapi/generator/kotlingen/ServerApiModuleGenerator.kt
package ru.tinkoff.kora.openapi.generator.kotlingen

import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.FileSpec
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.TypeSpec
import org.openapitools.codegen.model.OperationsMap
import ru.tinkoff.kora.openapi.generator.DelegateMethodBodyMode

class ServerApiModuleGenerator : AbstractKotlinGenerator<OperationsMap>() {
    override fun generate(ctx: OperationsMap): FileSpec {
        val b = TypeSpec.interfaceBuilder(ctx["classname"].toString() + "Module")
            .addAnnotation(generated())
            .addAnnotation(Classes.module.asKt())

        if (params.delegateMethodBodyMode != DelegateMethodBodyMode.NONE) {
            val delegateClass = ClassName(apiPackage, ctx["classname"].toString() + "Delegate");
            val o = TypeSpec.anonymousClassBuilder()
                .addAnnotation(generated())
                .addSuperinterface(delegateClass)
                .build()
            b.addFunction(FunSpec.builder("default" + ctx.get("classname") + "Delegate")
                .addStatement("return %L", o)
                .returns(delegateClass)
                .build());
        }

        return FileSpec.get(apiPackage, b.build())

    }
}
